sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -721,16 +728,9 @@ public static <T> TypedRead<T> read(SerializableFunction<SchemaAndRecord, T> par
         .setDatumReaderFactory(
             (SerializableFunction<TableSchema, AvroSource.DatumReaderFactory<T>>)
                 input -> {
-                  try {
-                    String jsonTableSchema = BigQueryIO.JSON_FACTORY.toString(input);
-                    return (AvroSource.DatumReaderFactory<T>)
-                        (writer, reader) ->
-                            new GenericDatumTransformer<>(parseFn, jsonTableSchema, writer);
-                  } catch (IOException e) {
-                    LOG.warn(
-                        String.format("Error while converting table schema %s to JSON!", input), e);
-                    return null;
-                  }
+                  TableSchema safeInput = checkStateNotNull(input);
+                  return (AvroSource.DatumReaderFactory<T>)
+                      (writer, reader) -> new GenericDatumTransformer<>(parseFn, safeInput, writer);
```

Comment:
This uses the new GenericDatumTransformer constructor that just takes the TableSchema input directly instead of needing to convert to JSON and then back again to the TableSchema later.